wasi-filesystem.wit.md

```diff
@@ -816,6 +797,54 @@ try-lock-exclusive: func() -> result<_, errno>
 unlock: func() -> result<_, errno>
 ```
 
+## `is-same-object`
+```wit
+/// Test whether two descriptors refer to the same filesystem object.
+///
+/// In POSIX, this corresponds to testing whether the two descriptors have the
+/// same device (`st_dev`) and inode (`st_ino` or `d_ino`) numbers.
+/// wasi-filesystem does not expose device and inode numbers, so this function
+/// may be used instead.
+is-same-object: func(other: descriptor) -> bool
+```
+
+## `metadata-hash`
+```wit
+/// Return a hash of the metadata associated with a filesystem object referred
+/// to by a descriptor.
```

Comment:
my honest impression is that it belongs to applications.
with the current description, i suspect an application will need to combine it with other properties like timestamp etc by itself anyway.
i feel it's simpler to provide `get-fileid-at`.

Reply:
It would indeed be simpler, but it'd have the problems discussed above: not all desirable impls have a meaningful fileid, and an exposed fileid leaks host information. Also, a fileid by itself isn't unique, and is often expected to be paired with a deviceid, but that would require a deviceid namespace. The nice thing about metadata-hash is that it avoids assigning any meaning to the number, so it can be computed in different ways depending on what the host implementation has.

Reply:
> Also, a fileid by itself isn't unique, and is often expected to be paired with a deviceid, but that would require a deviceid namespace.

I was interpeting the proposal as fileid being a combination of the device id and inode number.

> The nice thing about metadata-hash is that it avoids assigning any meaning to the number, so it can be computed in different ways depending on what the host implementation has.

A fileid could be a hash of everything that stays fixed for the lifetime of a file too, or assigned through any other consistent scheme.

Reply:
If the fileid is the device id and inode number, then how would it avoid the problems at the top of this issue?

If the fileid is a hash of everything that stays fixed for the lifetime of a file, how is that different from the `metadata-hash` I just proposed?

Reply:
> If the fileid is the device id and inode number, then how would it avoid the problems at the top of this issue?

Hashing them together combined with a secret value specified by the wasi runtime that is unknown to the wasm module (such that brute-forcing a map from hash to input is not possible) would avoid metadata leakage. As for some filesystems not having inodes, that is just as much an issue with metadata-hash as all other metadata can be non-unique between two different files. Also in both cases it would be possible to have the wasi runtime assign the id/hash based on eg the file path.

> If the fileid is a hash of everything that stays fixed for the lifetime of a file, how is that different from the metadata-hash I just proposed?

The metadata hash you specified includes the mtime, which can change without changing the identity of the file.

Reply:
I've now added documentation mentioning that a secret value known only to the implementation may be used.